In [1]:
# Importing the Dependencies
import numpy as np
import pandas as pd


from sklearn.metrics import accuracy_score
from sklearn.ensemble import IsolationForest
from sklearn.neighbors import LocalOutlierFactor

In [2]:
#loading datasets to panda dataframe
credit_card_test_data = pd.read_csv('E:\py programs\codsoft_ml\datasets\credit card\\fraudTest.csv')
credit_card_train_data = pd.read_csv('E:\py programs\codsoft_ml\datasets\credit card\\fraudTrain.csv')

credit_card_data = pd.concat([credit_card_train_data, credit_card_test_data], axis=0)

In [3]:
#duplicate values checking
credit_card_data.duplicated().sum()

0

We don't have any duplicate values after concatanating the two datasets

In [4]:
#first 5 rows
credit_card_data.head()

Unnamed: 0 trans_date_trans_time            cc_num  \
0           0   2019-01-01 00:00:18  2703186189652095   
1           1   2019-01-01 00:00:44      630423337322   
2           2   2019-01-01 00:00:51    38859492057661   
3           3   2019-01-01 00:01:16  3534093764340240   
4           4   2019-01-01 00:03:06   375534208663984   

                             merchant       category     amt      first  \
0          fraud_Rippin, Kub and Mann       misc_net    4.97   Jennifer   
1     fraud_Heller, Gutmann and Zieme    grocery_pos  107.23  Stephanie   
2                fraud_Lind-Buckridge  entertainment  220.11     Edward   
3  fraud_Kutch, Hermiston and Farrell  gas_transport   45.00     Jeremy   
4                 fraud_Keeling-Crist       misc_pos   41.96      Tyler   

      last gender                        street  ...      lat      long  \
0    Banks      F                561 Perry Cove  ...  36.0788  -81.1781   
1     Gill      F  43039 Riley Greens Suite 393  ...  48.8878 -118.2105   
2  Sanchez      M      594 White Dale Suite 530  ...  42.1808 -112.2620   
3    White      M   9443 Cynthia Court Apt. 038  ...  46.2306 -112.1138   
4   Garcia      M              408 Bradley Rest  ...  38.4207  -79.4629   

   city_pop                                job         dob  \
0      3495          Psychologist, counselling  1988-03-09   
1       149  Special educational needs teacher  1978-06-21   
2      4154        Nature conservation officer  1962-01-19   
3      1939                    Patent attorney  1967-01-12   
4        99     Dance movement psychotherapist  1986-03-28   

                          trans_num   unix_time  merch_lat  merch_long  \
0  0b242abb623afc578575680df30655b9  1325376018  36.011293  -82.048315   
1  1f76529f8574734946361c461b024d99  1325376044  49.159047 -118.186462   
2  a1a22d70485983eac12b5b88dad1cf95  1325376051  43.150704 -112.154481   
3  6b849c168bdad6f867558c3793159a81  1325376076  47.034331 -112.561071   
4  a41d7549acf90789359a9aa5346dcb46  1325376186  38.674999  -78.632459   

   is_fraud  
0         0  
1         0  
2         0  
3         0  
4         0  

[5 rows x 23 columns]

In [5]:
#last 5 rows
credit_card_data.tail()

Unnamed: 0 trans_date_trans_time            cc_num  \
555714      555714   2020-12-31 23:59:07    30560609640617   
555715      555715   2020-12-31 23:59:09  3556613125071656   
555716      555716   2020-12-31 23:59:15  6011724471098086   
555717      555717   2020-12-31 23:59:24     4079773899158   
555718      555718   2020-12-31 23:59:34  4170689372027579   

                     merchant        category     amt    first     last  \
555714  fraud_Reilly and Sons  health_fitness   43.77  Michael    Olson   
555715   fraud_Hoppe-Parisian       kids_pets  111.84     Jose  Vasquez   
555716        fraud_Rau-Robel       kids_pets   86.88      Ann   Lawson   
555717  fraud_Breitenberg LLC          travel    7.99     Eric  Preston   
555718      fraud_Dare-Marvin   entertainment   38.13   Samuel     Frey   

       gender                      street  ...      lat      long  city_pop  \
555714      M         558 Michael Estates  ...  40.4931  -91.8912       519   
555715      M         572 Davis Mountains  ...  29.0393  -95.4401     28739   
555716      F  144 Evans Islands Apt. 683  ...  46.1966 -118.9017      3684   
555717      M  7020 Doyle Stream Apt. 951  ...  44.6255 -116.4493       129   
555718      M    830 Myers Plaza Apt. 384  ...  35.6665  -97.4798    116001   

                   job         dob                         trans_num  \
555714    Town planner  1966-02-13  9b1f753c79894c9f4b71f04581835ada   
555715  Futures trader  1999-12-27  2090647dac2c89a1d86c514c427f5b91   
555716        Musician  1981-11-29  6c5b7c8add471975aa0fec023b2e8408   
555717    Cartographer  1965-12-15  14392d723bb7737606b2700ac791b7aa   
555718     Media buyer  1993-05-10  1765bb45b3aa3224b4cdcb6e7a96cee3   

         unix_time  merch_lat  merch_long  is_fraud  
555714  1388534347  39.946837  -91.333331         0  
555715  1388534349  29.661049  -96.186633         0  
555716  1388534355  46.658340 -119.715054         0  
555717  1388534364  44.470525 -117.080888         0  
555718  1388534374  36.210097  -97.036372         0  

[5 rows x 23 columns]

In [6]:
#data information
credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1852394 entries, 0 to 555718
Data columns (total 23 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   Unnamed: 0             int64  
 1   trans_date_trans_time  object 
 2   cc_num                 int64  
 3   merchant               object 
 4   category               object 
 5   amt                    float64
 6   first                  object 
 7   last                   object 
 8   gender                 object 
 9   street                 object 
 10  city                   object 
 11  state                  object 
 12  zip                    int64  
 13  lat                    float64
 14  long                   float64
 15  city_pop               int64  
 16  job                    object 
 17  dob                    object 
 18  trans_num              object 
 19  unix_time              int64  
 20  merch_lat              float64
 21  merch_long             float64
 22  is_fraud           

In [7]:
credit_card_data.dropna(inplace=True) #drops rows with null values

In [8]:
credit_card_data.isnull().sum()

Unnamed: 0               0
trans_date_trans_time    0
cc_num                   0
merchant                 0
category                 0
amt                      0
first                    0
last                     0
gender                   0
street                   0
city                     0
state                    0
zip                      0
lat                      0
long                     0
city_pop                 0
job                      0
dob                      0
trans_num                0
unix_time                0
merch_lat                0
merch_long               0
is_fraud                 0
dtype: int64

In [9]:
credit_card_data.isnull().values.any()

False

In [10]:
#no. of legit and fraudulent transactions in data
# 0 -> legit
#1 -> fraudulent
credit_card_data['is_fraud'].value_counts()

0    1842743
1       9651
Name: is_fraud, dtype: int64

In [11]:
## Get the Fraud and the normal dataset 

fraud = credit_card_data[credit_card_data['is_fraud']==1]

normal = credit_card_data[credit_card_data['is_fraud']==0]

In [12]:
print(fraud.shape,normal.shape)

(9651, 23) (1842743, 23)


We need to analyze more amount of information from the transaction data
How different are the amount of money used in different transaction classes?

In [13]:
fraud.amt.describe()

count    9651.000000
mean      530.661412
std       391.028873
min         1.060000
25%       240.075000
50%       390.000000
75%       902.365000
max      1376.040000
Name: amt, dtype: float64

In [14]:
normal.amt.describe()

count    1.842743e+06
mean     6.765128e+01
std      1.535481e+02
min      1.000000e+00
25%      9.610000e+00
50%      4.724000e+01
75%      8.256000e+01
max      2.894890e+04
Name: amt, dtype: float64

In [15]:
# Take some sample of the data

credit_card_data_sample= credit_card_data.sample(frac = 0.1,random_state=1)

credit_card_data_sample.shape

(185239, 23)

In [16]:
credit_card_data.shape

(1852394, 23)

In [17]:
#Determine the number of fraud and valid transactions in the dataset

Fraud = credit_card_data_sample[credit_card_data_sample['is_fraud']==1]

Valid = credit_card_data_sample[credit_card_data_sample['is_fraud']==0]

outlier_fraction = len(Fraud)/float(len(Valid))

In [18]:
print(outlier_fraction)

print("Fraud Cases : {}".format(len(Fraud)))

print("Valid Cases : {}".format(len(Valid)))

0.0052313118979785645
Fraud Cases : 964
Valid Cases : 184275


In [19]:
#Create independent and Dependent Features
columns = credit_card_data_sample.columns.tolist()
# Filter the columns to remove data we do not want 
columns = [c for c in columns if c not in ["is_fraud"]]
# Store the variable we are predicting 
target = "is_fraud"
# Define a random state 
state = np.random.RandomState(42)
X = credit_card_data_sample[columns]
Y = credit_card_data_sample[target]
X_outliers = state.uniform(low=0, high=1, size=(X.shape[0], X.shape[1]))
# Print the shapes of X & Y
print(X.shape)
print(Y.shape)

(185239, 22)
(185239,)


Isolation Forest Algorithm :
One of the newest techniques to detect anomalies is called Isolation Forests. The algorithm is based on the fact that anomalies are data points that are few and different. As a result of these properties, anomalies are susceptible to a mechanism called isolation.

This method is highly useful and is fundamentally different from all existing methods. It introduces the use of isolation as a more effective and efficient means to detect anomalies than the commonly used basic distance and density measures. Moreover, this method is an algorithm with a low linear time complexity and a small memory requirement. It builds a good performing model with a small number of trees using small sub-samples of fixed size, regardless of the size of a data set.

Typical machine learning methods tend to work better when the patterns they try to learn are balanced, meaning the same amount of good and bad behaviors are present in the dataset.

How Isolation Forests Work The Isolation Forest algorithm isolates observations by randomly selecting a feature and then randomly selecting a split value between the maximum and minimum values of the selected feature. The logic argument goes: isolating anomaly observations is easier because only a few conditions are needed to separate those cases from the normal observations. On the other hand, isolating normal observations require more conditions. Therefore, an anomaly score can be calculated as the number of conditions required to separate a given observation.

The way that the algorithm constructs the separation is by first creating isolation trees, or random decision trees. Then, the score is calculated as the path length to isolate the observation.


Local Outlier Factor(LOF) Algorithm
The LOF algorithm is an unsupervised outlier detection method which computes the local density deviation of a given data point with respect to its neighbors. It considers as outlier samples that have a substantially lower density than their neighbors.

The number of neighbors considered, (parameter n_neighbors) is typically chosen 1) greater than the minimum number of objects a cluster has to contain, so that other objects can be local outliers relative to this cluster, and 2) smaller than the maximum number of close by objects that can potentially be local outliers. In practice, such informations are generally not available, and taking n_neighbors=20 appears to work well in general.

In [20]:
#Define the outlier detection methods

##Define the outlier detection methods

classifiers = {
    "Isolation Forest":IsolationForest(n_estimators=100, max_samples=len(X), 
                                       contamination=outlier_fraction,random_state=state, verbose=0),
    "Local Outlier Factor":LocalOutlierFactor(n_neighbors=20, algorithm='auto', 
                                              leaf_size=30, metric='minkowski',
                                              p=2, metric_params=None, contamination=outlier_fraction),
}

type(classifiers)

dict

In [21]:
# We need to convert the strings to float otherwise we will get error hence we convert use labelencoder

from sklearn import preprocessing
le = preprocessing.LabelEncoder()
for column_name in X.columns:
    if X[column_name].dtype == object:
        X[column_name] = le.fit_transform(X[column_name])
    else:
        pass

C:\Users\USER\AppData\Local\Temp\ipykernel_5680\4269251600.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X[column_name] = le.fit_transform(X[column_name])
C:\Users\USER\AppData\Local\Temp\ipykernel_5680\4269251600.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X[column_name] = le.fit_transform(X[column_name])
C:\Users\USER\AppData\Local\Temp\ipykernel_5680\4269251600.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_in

In [22]:
n_outliers = len(Fraud)
for i, (clf_name,clf) in enumerate(classifiers.items()):
    #Fit the data and tag outliers
    if clf_name == "Local Outlier Factor":
        y_pred = clf.fit_predict(X)
        scores_prediction = clf.negative_outlier_factor_
    else:    
        clf.fit(X)
        scores_prediction = clf.decision_function(X)
        y_pred = clf.predict(X)
    #Reshape the prediction values to 0 for Valid transactions , 1 for Fraud transactions
    y_pred[y_pred == 1] = 0
    y_pred[y_pred == -1] = 1
    n_errors = (y_pred != Y).sum()
    # Run Classification Metrics
    print("{}: {}".format(clf_name,n_errors))
    print("Accuracy Score :")
    print(accuracy_score(Y,y_pred))

c:\Users\USER\anaconda3\Lib\site-packages\sklearn\base.py:439: UserWarning: X does not have valid feature names, but IsolationForest was fitted with feature names
  warnings.warn(


Isolation Forest: 1862
Accuracy Score :
0.9899481210760153
Local Outlier Factor: 1825
Accuracy Score :
0.9901478630310032


Observations :

Isolation Forest has a 98.99% which is less accurate than LOF of 99.01%.

When comparing error precision & recall for 2 models , the LOF performed much better than the Isolation Forest.